docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+
+##### Avoid unintentionally widening an inferred type with a type annotation
+
+Enforcing a wider type defeats the purpose of adding an explicit type declaration, as it _loses_ type information instead of adding it. Double-check that the declared type is narrower than the inferred type.
+
+🚫 Type declarations
+
+```typescript
+const name: string = 'METAMASK'; // Type 'string'
+
+const chainId: string = this.messagingSystem(
+  'NetworkController:getProviderConfig',
+).chainId; // Type 'string'
+
+const BUILT_IN_NETWORKS = new Map<string, `0x${string}`>([
+  ['mainnet', '0x1'],
+  ['sepolia', '0xaa36a7'],
+]); // Type 'Map<string, `0x${string}`>'
+```
+
+✅ Type inferences
+
+```typescript
+const name = 'METAMASK'; // Type 'METAMASK'
+
+const chainId = this.messagingSystem(
+  'NetworkController:getProviderConfig',
+).chainId; // Type '`0x${string}`'
+
+const BUILT_IN_NETWORKS = {
+  mainnet: '0x1',
+  sepolia: '0xaa36a7',
+} as const; // Type { readonly mainnet: '0x1'; readonly sepolia: '0xaa36a7'; }
+```
+
+```typescript
+type TransactionMeta = TransactionBase &
+  (
+    | {
+        status: Exclude<TransactionStatus, TransactionStatus.failed>;
+      }
+    | {
+        status: TransactionStatus.failed;
+        error: TransactionError;
+      }
+  );
+
+const updatedTransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected,
+};
+
+this.messagingSystem.publish(
+  `${controllerName}:transactionFinished`,
+  updatedTransactionMeta, // Expected type: 'TransactionMeta'
+);
+// Property 'error' is missing in type 'typeof updatedTransactionMeta' but required in type '{ status: TransactionStatus.failed; error: TransactionError; }'.ts(2345)
+```
+
+🚫 Add type annotation
+
+```typescript
+// Type 'TransactionMeta'
+const updatedTransactionMeta: TransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected,
+}; // resolves error
+```
+
+✅ Add `as const` and leave to inference
+
+```typescript
+// Type narrower than 'TransactionMeta': { status: TransactionStatus.rejected; ... }
+// Doesn't include 'error' property. Correctly narrowed between `TransactionMeta` discriminated union
+const updatedTransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected as const,
+}; // resolves error
+```
+
+### Type Annotations
+
+An explicit type annotation is acceptable for overriding an inferred type if...
+
+1. It can further narrow an inferred type, thus supplying type information that the compiler cannot infer or access.
+2. It is being used to enforce a type constraint rather than assign a type definition.
+
+##### Use the `satisfies` operator to enforce a type constraint or to validate the assigned type
+
+In the second case, an abstract, "narrowest supertype" is used to constrain or validate a type. TypeScript provides the `satisfies` operator for this use case. It is able to both enforce a type constraint and further narrow the assigned type through inference.
```

Comment:
Nit: The "In the second case" here caught me off guard the first time, I thought it was talking about the second example rather than the second case in the above section. Maybe this could be simplified:
```suggestion
TypeScript provides the `satisfies` operator for constraining or validating a type. It is able to both enforce a type constraint and further narrow the assigned type through inference.
```